L1Trigger/TrackFindingTMTT/interface/HTrphi.h

```diff
@@ -0,0 +1,166 @@
+
+    // Define the order in which the hardware processes rows of the HT array when it outputs track candidates.
+    // Currently corresponds to highest Pt tracks first.
+    // If two tracks have the same Pt, the -ve charge one is output before the +ve charge one.
+    vector<unsigned int> rowOrder(unsigned int numRows) const;
+
+  private:
+    float invPtToDphi_;  // conversion constant.
+    unsigned int shape_;
+    std::vector<std::vector<std::pair<float, float> > > cellCenters_;
+
+    //--- Specifications of HT array.
+
+    float maxAbsQoverPtAxis_;        // Max. |q/Pt| covered by  HT array.
+    unsigned int nBinsQoverPtAxis_;  // Number of bins in HT array in q/Pt.
+    float binSizeQoverPtAxis_;       // HT array bin size in q/Pt.
+
+    float chosenRofPhi_;            // Use phi of track at radius="chosenRofPhi" to define one of the r-phi HT axes.
+    float phiCentreSector_;         // phiTrk angle of centre of this (eta,phi) sector.
+    float maxAbsPhiTrkAxis_;        // Half-width of phiTrk axis in HT array.
+    unsigned int nBinsPhiTrkAxis_;  // Number of bins in HT array in phiTrk axis.
+    float binSizePhiTrkAxis_;       // HT array bin size in phiTrk
+
+    bool enableMerge2x2_;  // Optionally merge 2x2 neighbouring cells into a single cell at low Pt, to reduce efficiency loss due to scattering. (Used also by mini-HT).
+    float minInvPtToMerge2x2_;
+
+    //--- Options when filling HT array.
+
+    unsigned int
+        killSomeHTCellsRphi_;  // Take all cells in HT array crossed by line corresponding to each stub (= 0) or take only some to reduce rate at cost of efficiency ( > 0)
+    bool handleStripsRphiHT_;  // Should algorithm allow for uncertainty in stub (r,z) coordinate caused by length of 2S module strips when fill stubs in r-phi HT?
+    // Options for killing stubs/tracks that cant be sent within time-multiplexed period.
+    bool busyInputSectorKill_;
+    unsigned int busyInputSectorNumStubs_;
+    bool busySectorKill_;
+    unsigned int busySectorNumStubs_;
+    vector<unsigned int> busySectorMbinRanges_;
+    bool busySectorUseMbinRanges_;
+    vector<unsigned int> busySectorMbinOrder_;
+    bool busySectorUseMbinOrder_;
+    vector<unsigned int> busySectorMbinLow_;
+    vector<unsigned int> busySectorMbinHigh_;
+
+    //--- Checks that stub filling is compatible with limitations of firmware.
+
+    // Maximum |gradient| of line corresponding to any stub. Should be less than the value of 1.0 assumed by the firmware.
+    static float maxLineGradient_;
+    // Error count when stub added to cell which does not lie NE, E or SE of stub added to previous HT column.
+    static unsigned int numErrorsTypeA_;
+    // Error count when stub added to more than 2 cells in one HT column (problem only for Thomas' firmware).
+    static unsigned int numErrorsTypeB_;
+    // Error count normalisation
+    static unsigned int numErrorsNormalisation_;
```

Comment:
I'm now using a mix of "static thread_local" and "std::atomic".

Reply:
The `thread_local`s should be avoided as well. Could you explain why a global state would be needed here, i.e. why all these can not be either local variables or normal member variables?

Reply:
The HTrphi object is reinstantiated for each new event (to avoid the risk of data from a previous event remaining). These 4 static variables are used to monitor the HTrphi behaviour over the entire run, with these values printed out in the endJob() by Histos.cc. If the HTrphi is optimally configured, the first should be approximately 1.0, and the other three 0.0. If they differ from this, the amount they differ indicates how the HTrphi should be reconfigured to solve the problem.

If you really don't like these statics (which now use std::atomic, so should be safe), the only alternative would be make these 4 variables non-static, extract their values at the end of each event in the EDProducer, and pass them to the Histogramming class, which could then monitor their values over the entire run. The disadvantage of doing this is that TMTrackProducer.cc would need to receive/transmit these 4 variables, adding slightly to its complexity.

Reply:
Sometimes, when the solution to a problem introduces a lot of complexity and unwanted effects, it's worth asking: does the problem really need to be solved?

In this case, do we really need to monitor at this level of granularity in production code?

For all of these cases in this PR, I would propose that fine-grained monitoring for debugging should be accomplished by *optional* event products that do not affect production, but can be enabled by special runs. These products can then be consumed and analyzed by dedicated validation code in the proper subsystem, which is how the rest of CMSSW works. I took this approach e.g. in the calorimeter digitization, and found that it worked quite well: https://github.com/cms-sw/cmssw/blob/f42c909c15e3a088d9ef3f8b535a181a7509c4ea/SimCalorimetry/HcalSimProducers/src/HcalDigiProducer.cc#L17-L19

Reply:
@tomalin Thanks for the explanation. The problem with `static std::atomic<T>` is that while it is thread safe, it is a ticking time bomb: the use of `HTrphi` is fixed to a single EDModule. That may be sufficient today, but what if tomorrow (or next year, or five years from now) someone wants to run e.g. the same EDModule with a bit different configuration in the same job. Then these global counts would not be the expected ones.

The best way to do aggregation over events is via objects owned by the EDModule that have the proper life time.

By the way, with "the entire run", do you mean the duration of the job, or the duration of a `Run`?

Reply:
Entire job

Reply:
These statics have been now become non-static data members of the EDProducer.